File: usercorn/arch/x86/cgc.py
from unicorn.x86_const import *
import os
import struct
import sys

def read_fdset(cls, addr):
    n, = struct.unpack('<I', cls.mem_read(addr, 4))
    fds = cls.mem_read(addr + 4, n * 4)
    return struct.unpack('<%dI' % n, fds)

def write_fdset(cls, addr, fds):
    mem = struct.pack('<I%dI' % len(fds), len(fds), *fds)
    cls.mem_write(addr, mem)

def read_timeval(cls, addr):
    sec, us = struct.unpack('<II', cls.mem_read(addr, 8))
    return sec + us / 1000000.0

def syscall(cls):
    regs = [UC_X86_REG_EAX, UC_X86_REG_EBX, UC_X86_REG_ECX, UC_X86_REG_EDX, UC_X86_REG_ESI, UC_X86_REG_EDI, UC_X86_REG_EBP]
    num, a1, a2, a3, a4, a5, a6 = [cls.reg_read(r) for r in regs]
    ret = 0
    # _terminate
    if num == 1:
        sys.exit(a1)
    # transmit
    elif num == 2:
        n = os.write(a1, cls.mem_read(a2, a3))
        cls.mem_write(a4, cls.pack_addr(n))
    # receive
    elif num == 3: 
        tmp = os.read(a1, a3)
        cls.mem_write(a2, tmp)
        cls.mem_write(a4, cls.pack_addr(len(tmp)))
    # fdwait
    elif num == 4:
        readfds = read_fdset(cls, a2)
        writefds = read_fdset(cls, a3)
        timeout = read_timeval(cls, a4)
        r, w, _ = select.select(readfds, writefds, [], timeout)
        write_fdset(cls, a5, r + w)
    # allocate
    elif num == 5:
        addr = cls.mmap(a1)
        cls.mem_write(a3, cls.pack_addr(addr))
    # deallocate
    elif num == 6:
        pass
    # random
    elif num == 7: 
        rnd = os.urandom(a2)
        cls.mem_write(a1, rnd)
        cls.mem_write(a3, cls.pack_addr(a2))
    else:
        print 'Unsupported syscall:', num
        sys.exit(1)
    cls.reg_write(UC_X86_REG_EAX, ret)

def interrupt(cls, intno):
    if intno == 0x80:
        syscall(cls)
    else:
        raise NotImplementedError('unhandled interrupt %d' % intno)

File: go/arch/x86/cgc.go
package x86

import (
	"crypto/rand"
	"github.com/pkg/errors"
	uc "github.com/unicorn-engine/unicorn/bindings/go/unicorn"
	"syscall"

	co "github.com/lunixbochs/usercorn/go/kernel/common"
	"github.com/lunixbochs/usercorn/go/models"
	"github.com/lunixbochs/usercorn/go/models/cpu"
	"github.com/lunixbochs/usercorn/go/native"
)

var cgcSysNum = map[int]string{
	1: "_terminate",
	2: "transmit",
	3: "receive",
	4: "fdwait",
	5: "allocate",
	6: "deallocate",
	7: "random",
}

type CgcKernel struct {
	*co.KernelBase
}

func (k *CgcKernel) Literal_terminate(code int) {
	k.U.Exit(models.ExitStatus(code))
}

func (k *CgcKernel) Transmit(fd co.Fd, buf co.Buf, size co.Len, ret co.Obuf) int {
	if fd == 0 {
		fd = 1
	}
	mem, err := k.U.MemRead(buf.Addr, uint64(size))
	if err != nil {
		return -1 // FIXME
	}
	n, err := syscall.Write(int(fd), mem)
	if err != nil {
		return -1 // FIXME
	}
	if ret.Addr > 0 {
		if err := ret.Pack(int32(n)); err != nil {
			return -1 // FIXME
		}
	}
	return 0
}

func (k *CgcKernel) Receive(fd co.Fd, buf co.Obuf, size co.Len, ret co.Obuf) int {
	tmp := make([]byte, size)
	n, err := syscall.Read(int(fd), tmp)
	if err != nil {
		return -1 // FIXME
	}
	if err := buf.Pack(tmp[:n]); err != nil {
		return -1 // FIXME
	}
	if ret.Addr > 0 {
		if err := ret.Pack(int32(n)); err != nil {
			return -1 // FIXME
		}
	}
	return 0
}

func (k *CgcKernel) Fdwait(nfds int, reads, writes, timeoutBuf co.Buf, readyFds co.Obuf) int {
	var readSet, writeSet *native.Fdset32
	var timeout native.Timespec
	reads.Unpack(&readSet)
	writes.Unpack(&writeSet)
	timeoutBuf.Unpack(&timeout)

	readNative := readSet.Native()
	writeNative := writeSet.Native()

	n, err := native.Select(nfds, readNative, writeNative, &timeout)
	if err != nil {
		return -1 // FIXME?
	} else {
		readyFds.Pack(int32(n))
	}
	return 0
}

func (k *CgcKernel) Allocate(size uint32, executable int32, ret co.Obuf) int {
	// round up to nearest page
	size = (size + 0x1000) & ^uint32(0x1000-1)
	addr, _ := k.U.Malloc(uint64(size))
	if executable != 0 {
		k.U.MemProt(addr, uint64(size), uc.PROT_ALL)
	}
	ret.Pack(uint32(addr))
	return 0
}

func (k *CgcKernel) Deallocate(addr, size uint32) int {
	// addr must be multiple of page size
	if addr&(0x1000-1) > 0 {
		return -1 // FIXME
	}
	// all pages containing a part of the range are unmapped
	var pages []*models.Mmap
	for _, mmap := range k.U.Mappings() {
		// does addr overlap mapping?
		if (uint64(addr) >= mmap.Addr && uint64(addr) < mmap.Addr+mmap.Size) ||
			(uint64(addr) < mmap.Addr && uint64(addr+size) > mmap.Addr) {
			pages = append(pages, mmap)
		}
	}
	for _, page := range pages {
		k.U.MemUnmap(page.Addr, page.Size)
	}
	return 0
}

func (k *CgcKernel) Random(buf co.Obuf, size uint32, ret co.Obuf) {
	tmp := make([]byte, size)
	n, _ := rand.Read(tmp)
	tmp = tmp[:n]
	buf.Pack(tmp)
	ret.Pack(uint32(n))
}

func CgcInit(u models.Usercorn, args, env []string) error {
	// TODO: does CGC even specify argv?
	// TODO: also, I seem to remember something about mapping in 16kb of random data
	secretPage := uint64(0x4347c000)
	if err := u.MemMap(secretPage, 0x1000, cpu.PROT_READ); err != nil {
		return err
	}
	tmp := make([]byte, 0x1000)
	if _, err := rand.Read(tmp); err != nil {
		return errors.Wrap(err, "rand.Read() failed")
	}
	if err := u.MemWrite(secretPage, tmp); err != nil {
		return err
	}
	u.RegWrite(uc.X86_REG_ECX, secretPage)
	return u.MapStack(0xbaaab000, 0x800000, false)
}

func CgcSyscall(u models.Usercorn) {
	eax, _ := u.RegRead(uc.X86_REG_EAX)
	name, _ := cgcSysNum[int(eax)]
	ret, _ := u.Syscall(int(eax), name, co.RegArgs(u, LinuxRegs))
	u.RegWrite(uc.X86_REG_EAX, ret)
}

func CgcInterrupt(u models.Usercorn, intno uint32) {
	if intno == 0x80 {
		CgcSyscall(u)
	}
}

func CgcKernels(u models.Usercorn) []interface{} {
	kernel := &CgcKernel{&co.KernelBase{}}
	return []interface{}{kernel}
}

func init() {
	Arch.RegisterOS(&models.OS{
		Name:      "cgc",
		Init:      CgcInit,
		Interrupt: CgcInterrupt,
		Kernels:   CgcKernels,
	})
}
